File: jetbrains/plugin/src/main/kotlin/ai/kilocode/jetbrains/tools/CreateDraftTool.kt
// SPDX-FileCopyrightText: 2025 Weibo, Inc.
//
// SPDX-License-Identifier: Apache-2.0

package ai.kilocode.jetbrains.tools

import ai.kilocode.jetbrains.filesystem.DraftFileSystemProvider
import com.intellij.openapi.diagnostic.Logger

/**
 * Tool handler for create_draft tool.
 * Creates ephemeral in-memory draft documents that appear as editor tabs.
 */
class CreateDraftTool(private val project: com.intellij.openapi.project.Project) {
    private val logger = Logger.getInstance(CreateDraftTool::class.java)

    /**
     * Handle create_draft tool invocation.
     * @param params Tool parameters containing title and content
     * @return Result map with success message and draft path
     */
    fun handle(params: Map<String, Any?>): Map<String, Any?> {
        try {
            // Extract and validate parameters
            val title = params["title"] as? String
            val content = params["content"] as? String

            if (title.isNullOrBlank()) {
                return mapOf(
                    "result" to "Error: 'title' parameter is required",
                    "error" to "Missing required parameter: title",
                )
            }

            if (content == null) {
                return mapOf(
                    "result" to "Error: 'content' parameter is required",
                    "error" to "Missing required parameter: content",
                )
            }

            // Create draft document
            val draftProvider = DraftFileSystemProvider.getInstance()
            val draftPath = draftProvider.createAndOpen(title, content, project)

            logger.info("Created draft document: $draftPath")

            // Return success message with instructions
            val message = """Created draft document "$title". The document has been opened in an editor tab.

You can now:
- Read it using: read_file with path "$draftPath"
- Update it using: write_to_file with path "$draftPath"

The document will be discarded when the editor session ends."""

            return mapOf(
                "result" to message,
                "id" to "create_draft",
                "draftPath" to draftPath,
            )
        } catch (e: Exception) {
            logger.error("Error creating draft document", e)
            return mapOf(
                "result" to "Error creating draft document: ${e.message}",
                "error" to e.message ?: "Unknown error",
            )
        }
    }
}

File: src/core/tools/CreateDraftTool.ts
// kilocode_change - new file
import { Task } from "../task/Task"
import { formatResponse } from "../prompts/responses"
import { getDraftFileSystem } from "../../services/planning"
import { BaseTool, ToolCallbacks } from "./BaseTool"
import type { ToolUse } from "../../shared/tools"

interface CreateDraftParams {
	title: string
	content: string
}

export class CreateDraftTool extends BaseTool<"create_draft"> {
	readonly name = "create_draft" as const

	parseLegacy(params: Partial<Record<string, string>>): CreateDraftParams {
		return {
			title: params.title || "",
			content: params.content || "",
		}
	}

	async execute(params: CreateDraftParams, task: Task, callbacks: ToolCallbacks): Promise<void> {
		const { title, content } = params
		const { handleError, pushToolResult } = callbacks

		// Validate required parameters
		if (!title) {
			task.consecutiveMistakeCount++
			task.recordToolError("create_draft")
			pushToolResult(await task.sayAndCreateMissingParamError("create_draft", "title"))
			return
		}

		if (content === undefined || content === null) {
			task.consecutiveMistakeCount++
			task.recordToolError("create_draft")
			pushToolResult(await task.sayAndCreateMissingParamError("create_draft", "content"))
			return
		}

		// Validate title length
		if (title.length > 255) {
			task.consecutiveMistakeCount++
			task.recordToolError("create_draft")
			pushToolResult(formatResponse.toolError("Title must be 255 characters or less"))
			return
		}

		// Validate content is not too large (prevent memory issues)
		if (content.length > 1000000) {
			task.consecutiveMistakeCount++
			task.recordToolError("create_draft")
			pushToolResult(formatResponse.toolError("Content must be 1MB or less"))
			return
		}

		task.consecutiveMistakeCount = 0

		try {
			const fs = getDraftFileSystem()
			const draftPath = await fs.createAndOpen(title, content)

			// Return success message with instructions
			const message = `Created draft document "${title}". The document has been opened in an editor tab.\n\nYou can now:\n- Read it using: read_file with path "${draftPath}"\n- Update it using: write_to_file with path "${draftPath}"\n\nThe document will be discarded when the editor session ends.`

			pushToolResult(formatResponse.toolResult(message))
			task.recordToolUsage("create_draft")
		} catch (error) {
			await handleError("creating draft document", error as Error)
		}
	}

	override async handlePartial(task: Task, block: ToolUse<"create_draft">): Promise<void> {
		// Show "Creating draft..." message during streaming
		const title = this.removeClosingTag("title", block.params.title, block.partial)
		const content = this.removeClosingTag("content", block.params.content, block.partial)

		if (title) {
			const partialMessage = JSON.stringify({
				tool: "createDraft",
				title: title,
				content: content,
			})

			await task.ask("tool", partialMessage, block.partial).catch(() => {})
		}
	}
}

export const createDraftTool = new CreateDraftTool()
